dittofs-operator/internal/controller/dittoserver_controller.go

```diff
@@ -0,0 +1,329 @@
+			return err
+		}
+
+		serviceType := corev1.ServiceTypeClusterIP
+		if dittoServer.Spec.Service.Type != "" {
+			serviceType = corev1.ServiceType(dittoServer.Spec.Service.Type)
+		}
+
+		service.Spec = corev1.ServiceSpec{
+			Type: serviceType,
+			Selector: map[string]string{
+				"app":      "dittofs-server",
+				"instance": dittoServer.Name,
+			},
+			Ports: []corev1.ServicePort{
+				{
+					Name:     "nfs",
+					Port:     2049,
+					Protocol: corev1.ProtocolTCP,
+				},
+				{
+					Name:     "metrics",
+					Port:     9090,
+					Protocol: corev1.ProtocolTCP,
+				},
+			},
+		}
+
+		if dittoServer.Spec.Service.Annotations != nil {
+			service.Annotations = dittoServer.Spec.Service.Annotations
+		}
+
+		return nil
+	})
+
+	return err
+}
+
+func (r *DittoServerReconciler) reconcileStatefulSet(ctx context.Context, dittoServer *dittoiov1alpha1.DittoServer, replicas int32) error {
+	statefulSet := &appsv1.StatefulSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      dittoServer.Name,
+			Namespace: dittoServer.Namespace,
+		},
+	}
+
+	_, err := controllerutil.CreateOrUpdate(ctx, r.Client, statefulSet, func() error {
+		if err := controllerutil.SetControllerReference(dittoServer, statefulSet, r.Scheme); err != nil {
+			return err
+		}
+
+		labels := map[string]string{
+			"app":      "dittofs-server",
+			"instance": dittoServer.Name,
+		}
+
+		volumeMounts := []corev1.VolumeMount{
+			{
+				Name:      "metadata",
+				MountPath: "/data/metadata",
+			},
+			{
+				Name:      "config",
+				MountPath: "/config",
+			},
+		}
+
+		if dittoServer.Spec.Storage.ContentSize != "" {
+			volumeMounts = append(volumeMounts, corev1.VolumeMount{
+				Name:      "content",
+				MountPath: "/data/content",
+			})
+		}
+
+		volumeClaimTemplates := []corev1.PersistentVolumeClaim{
+			{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "metadata",
+				},
+				Spec: corev1.PersistentVolumeClaimSpec{
+					AccessModes: []corev1.PersistentVolumeAccessMode{
+						corev1.ReadWriteOnce,
+					},
+					Resources: corev1.VolumeResourceRequirements{
+						Requests: corev1.ResourceList{
+							corev1.ResourceStorage: resource.MustParse(dittoServer.Spec.Storage.MetadataSize),
+						},
+					},
+				},
+			},
+		}
+
+		if dittoServer.Spec.Storage.ContentSize != "" {
+			volumeClaimTemplates = append(volumeClaimTemplates, corev1.PersistentVolumeClaim{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "content",
+				},
+				Spec: corev1.PersistentVolumeClaimSpec{
+					AccessModes: []corev1.PersistentVolumeAccessMode{
+						corev1.ReadWriteOnce,
+					},
+					Resources: corev1.VolumeResourceRequirements{
+						Requests: corev1.ResourceList{
+							corev1.ResourceStorage: resource.MustParse(dittoServer.Spec.Storage.ContentSize),
+						},
+					},
+				},
+			})
+		}
+
+		if dittoServer.Spec.Storage.StorageClassName != nil {
+			for i := range volumeClaimTemplates {
+				volumeClaimTemplates[i].Spec.StorageClassName = dittoServer.Spec.Storage.StorageClassName
```

Comment:
We could set this directly during the PVC creation, so we avoid a post-processing loop